File: workstation/src/test/java/io/xj/model/meme/MemeTaxonomyTest.java
package io.xj.model.meme;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 TemplateConfig has Meme categories
 https://github.com/xjmusic/workstation/issues/209
 <p>
 A template configuration has a field called `memeTaxonomy` which defines the taxonomy of memes.
 <p>
 For example, this might look like
 <p>
 ```
 memeTaxonomy=CITY[CHICAGO,DENVER,PHILADELPHIA]
 ```
 <p>
 That would tell XJ about the existence of a meme category called City with values `CHICAGO`, `DENVER`, and `PHILADELPHIA`. And these would function as exclusion like numeric memes, e.g. after content having `CHICAGO` is chosen, we can choose nothing with `DENVER` or `PHILADELPHIA`.
 */
public class MemeTaxonomyTest {
  @Test
  public void testDefaultCategoryName() {
    var subject = MemeTaxonomy.fromList(List.of(Map.of("memes", List.of("WINTER", "SPRING", "SUMMER", "FALL"))));

    assertEquals("CATEGORY", subject.getCategories().get(0).getName());
  }

  @Test
  public void test_fromString_toString() {
    var subject = MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

    assertEquals("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
  }

  @Test
  public void test_fromListOfMaps_toListOfMaps() {
    var subject = MemeTaxonomy.fromList(List.of(
      Map.of(
        "name", "CO111LOR",
        "memes", List.of("RED ", "GR333EEN", "BLUE#@")
      ),
      Map.of(
        "name", "SIZ$%@E",
        "memes", List.of("L44A&*(RGE", "MED$IUM", "SMA)(&&LL")
      )));

    assertEquals(List.of(
      Map.of(
        "name", "COLOR",
        "memes", List.of("RED", "GREEN", "BLUE")
      ),
      Map.of(
        "name", "SIZE",
        "memes", List.of("LARGE", "MEDIUM", "SMALL")
      )
    ), subject.toList());
  }

  @Test
  public void testStripNonAlphabetical() {
    var subject = MemeTaxonomy.fromString("COLOR [RED, GREEN, BLUE];    SIZE [LARGE, MEDIUM, SMALL ]");

    assertEquals("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
  }

  @Test
  public void getCategories() {
    var subject = MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

    assertEquals(2, subject.getCategories().size());
    assertEquals("COLOR", subject.getCategories().get(0).getName());
    assertEquals("RED", subject.getCategories().get(0).getMemes().get(0));
    assertEquals("GREEN", subject.getCategories().get(0).getMemes().get(1));
    assertEquals("BLUE", subject.getCategories().get(0).getMemes().get(2));
    assertEquals("SIZE", subject.getCategories().get(1).getName());
    assertEquals("LARGE", subject.getCategories().get(1).getMemes().get(0));
    assertEquals("MEDIUM", subject.getCategories().get(1).getMemes().get(1));
    assertEquals("SMALL", subject.getCategories().get(1).getMemes().get(2));
  }

  @Test
  public void isAllowed() {
    assertTrue(MemeTaxonomy.fromString("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed(List.of("PEACHES")));
    assertTrue(MemeTaxonomy.fromString("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed(List.of("DENVER")));
    assertFalse(MemeTaxonomy.fromString("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed(List.of("DENVER", "PHILADELPHIA")));
    assertTrue(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("RED", "LARGE")));
    assertTrue(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("GREEN", "MEDIUM", "PEACHES")));
    assertFalse(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("RED", "BLUE", "LARGE")));
    assertFalse(MemeTaxonomy.fromString("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed(List.of("GREEN", "MEDIUM", "PEACHES", "SMALL")));
  }

  @Test
  public void isAllowed_alreadyPresentFromTaxonomy() {
    assertTrue(MemeTaxonomy.fromString("CITY[ABERDEEN,NAGOYA]").isAllowed(List.of("ABERDEEN", "ABERDEEN")));
  }


}

File: engine/test/meme/MemeTaxonomyTest.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <gtest/gtest.h>
#include <set>

#include "xjmusic/meme/MemeTaxonomy.h"

using namespace XJ;


/**
 TemplateConfig has Meme categories
 https://github.com/xjmusic/xjmusic/issues/209
 <p>
 A template configuration has a field called `memeTaxonomy` which defines the taxonomy of memes.
 <p>
 For example, this might look like
 <p>
 ```
 memeTaxonomy=CITY[CHICAGO,DENVER,PHILADELPHIA]
 ```
 <p>
 That would tell XJ about the existence of a meme category called City with values `CHICAGO`, `DENVER`, and `PHILADELPHIA`. And these would function as exclusion like numeric memes, e.g. after content having `CHICAGO` is chosen, we can choose nothing with `DENVER` or `PHILADELPHIA`.
 */

static bool setContains(const std::set<std::string> &items, const char *string) {
  return std::any_of(items.begin(), items.end(), [string](const std::string &item) {
    return item == string;
  });
}

TEST(MemeTaxonomyTest, DefaultCategoryName) {
  std::set<MapStringToOneOrManyString> input = {
      {{"memes", std::set<std::string>{"WINTER", "SPRING", "SUMMER", "FALL"}}},
  };

  auto subject = MemeTaxonomy::fromSet(input);

  EXPECT_EQ("CATEGORY", subject.getCategories().begin()->getName());
}

TEST(MemeTaxonomyTest, CategoryHasMemes) {
  std::set<MapStringToOneOrManyString> input = {
      {{"memes", std::set<std::string>{"WINTER", "SPRING", "SUMMER", "FALL"}}},
  };

  MemeTaxonomy subject(input);

  ASSERT_TRUE(subject.getCategories().begin()->hasMemes());
}

TEST(MemeTaxonomyTest, FromStringToString) {
  const MemeTaxonomy subject("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

  ASSERT_EQ("COLOR[BLUE,GREEN,RED];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
}

TEST(MemeTaxonomyTest, FromSetToList) {
  std::set<MapStringToOneOrManyString> input = {
      {{"name", "CO111LOR"}, {"memes", std::set<std::string>{"RED ", "GR333EEN", "BLUE#@"}}},
      {{"name", "SIZ$%@E"}, {"memes", std::set<std::string>{"L44A&*(RGE", "MED$IUM", "SMA)(&&LL"}}}};

  const MemeTaxonomy subject(input);

  const std::set<MapStringToOneOrManyString> expected = {
      {{"name", "COLOR"}, {"memes", std::set<std::string>{"RED", "GREEN", "BLUE"}}},
      {{"name", "SIZE"}, {"memes", std::set<std::string>{"LARGE", "MEDIUM", "SMALL"}}}
  };

  ASSERT_EQ(expected, subject.toList());
}

TEST(MemeTaxonomyTest, TestStripNonAlphabetical) {
  const MemeTaxonomy subject("COLOR [RED, GREEN, BLUE];    SIZE [LARGE, MEDIUM, SMALL ]");

  ASSERT_EQ("COLOR[BLUE,GREEN,RED];SIZE[LARGE,MEDIUM,SMALL]", subject.toString());
}

TEST(MemeTaxonomyTest, GetCategories) {
  MemeTaxonomy subject("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]");

  std::vector<MemeCategory> categories;
  for (const auto &category: subject.getCategories()) {
    categories.push_back(category);
  }
  std::sort(categories.begin(), categories.end());

  ASSERT_EQ(2, categories.size());
  ASSERT_EQ("COLOR", categories[0].getName());
  ASSERT_TRUE(setContains(categories[0].getMemes(), "RED"));
  ASSERT_TRUE(setContains(categories[0].getMemes(), "GREEN"));
  ASSERT_TRUE(setContains(categories[0].getMemes(), "BLUE"));
  ASSERT_EQ("SIZE", categories[1].getName());
  ASSERT_TRUE(setContains(categories[1].getMemes(), "LARGE"));
  ASSERT_TRUE(setContains(categories[1].getMemes(), "MEDIUM"));
  ASSERT_TRUE(setContains(categories[1].getMemes(), "SMALL"));
}

TEST(MemeTaxonomyTest, IsAllowed) {
  ASSERT_TRUE(MemeTaxonomy("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed({"PEACHES"}));
  ASSERT_TRUE(MemeTaxonomy("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed({"DENVER"}));
  ASSERT_FALSE(MemeTaxonomy("CITY[CHICAGO,DENVER,PHILADELPHIA]").isAllowed({"DENVER", "PHILADELPHIA"}));
  ASSERT_TRUE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"RED", "LARGE"}));
  ASSERT_TRUE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"GREEN", "MEDIUM", "PEACHES"}));
  ASSERT_FALSE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"RED", "BLUE", "LARGE"}));
  ASSERT_FALSE(MemeTaxonomy("COLOR[RED,GREEN,BLUE];SIZE[LARGE,MEDIUM,SMALL]").isAllowed({"GREEN", "MEDIUM", "PEACHES", "SMALL"}));
}

TEST(MemeTaxonomyTest, IsAllowed_alreadyPresentFromTaxonomy) {
  ASSERT_TRUE(MemeTaxonomy("CITY[ABERDEEN,NAGOYA]").isAllowed({"ABERDEEN", "ABERDEEN"}));
}

